src/hazardous/secret_stream/xchacha20poly1305.rs

```diff
@@ -0,0 +1,455 @@
+// MIT License
+
+// Copyright (c) 2018-2019 The orion Developers
+
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+// Based on the implementation used by libsodium
+// https://download.libsodium.org/doc/secret-key_cryptography/secretstream
+// https://github.com/jedisct1/libsodium/blob/stable/src/libsodium/crypto_secretstream/xchacha20poly1305/secretstream_xchacha20poly1305.c
+
+use crate::endianness::store_u64_into_le;
+use crate::errors::UnknownCryptoError;
+use crate::hazardous::mac::poly1305::{init, OneTimeKey, POLY1305_KEYSIZE, POLY1305_OUTSIZE};
+use crate::hazardous::stream::chacha20::{
+	encrypt as chacha20_enc, encrypt_in_place as chacha20_enc_in_place, hchacha20,
+	Nonce as chacha20Nonce, SecretKey as chacha20Key, CHACHA_KEYSIZE, HCHACHA_NONCESIZE,
+	IETF_CHACHA_NONCESIZE,
+};
+use crate::util::secure_cmp;
+use bitflags::bitflags;
+
+//TODO Add more tests
+
+//TODO: Im not sure if this is the right place for the bit flags and if they are needed at all.
+// It would also be possible to only define the REKEY value and pass the tag around as u8 and define
+// the user friendly bitflags/enum in the safe API. Im not sure how to work around using the
+// bitflags create. rust enums are rather limited in functionality as flags and im not a fan of
+// passing around raw u8 bitflags.
```

Comment:
I think this is the right place to put it. In regards to using the bitflags crate, I think we can keep it for now. What is really nice is that it has no `unsafe` usage and is used in a number of other crates as well.